[WIP] Phase 1 of the Implementation Plan

ekiru opened the pull request:
From #1 

* <del>Create an `ActivityMon::Record` class with `self.abstract_class = true` and `self.table_name_prefix = 'activitymon'` from which to inherit all the other ActivityMon models with their own tables.</del>
  * This turns out to be unnecessary: we can just set the prefix on the `ActivityMon` module
  * [x] Define `ActivityMon.table_name_prefix` to "activitymon"
    * Done in https://github.com/kibimon/monstodon/pull/3/commits/9baea84792d03d828da1e1598ab0067e41ce7272#diff-716c55d26e9f82047a4b253114b75d3f
* [x] Create an `ActivityMon::Species` model with just `id`, `name`, and `uri` columns.
  * Done in https://github.com/kibimon/monstodon/pull/3/commits/9baea84792d03d828da1e1598ab0067e41ce7272
* [x] Create an `ActivityMon::Trainer` model inheriting from `Account`.
  * Done in https://github.com/kibimon/monstodon/pull/3/commits/b8eb74474dedc85d1ef1b0d469d614ff61bebcd7#diff-4c9c49aa598af63e3db7747eb0a72f47
  * [x] Create a migration which adds a `type` column defaulting to `ActivityMon::Trainer` to the `accounts` table, defaulting all existing accounts to `ActivityMon::Trainer`.
    * Done in https://github.com/kibimon/monstodon/pull/3/commits/b8eb74474dedc85d1ef1b0d469d614ff61bebcd7#diff-54a532b490bf122cba18ed60b9062c5e
  * [x] <ins>Create a migration to backfill the `type` column on the `accounts` table.</ins>
    * Done in https://github.com/kibimon/monstodon/pull/3/commits/1a93edf150dd8ec81e7a12bc3c268754992079a0
  * [x] Add a validation to ensure that Trainers do not have owners or species.
    * Done in https://github.com/kibimon/monstodon/pull/3/commits/65f2c89a48c545f0dd665437e9a28bb334cdf393#diff-4c9c49aa598af63e3db7747eb0a72f47R53
  * [x] ??? Make sure any existing accounts stuff that needs to be is aware of Trainer.
    * Done in https://github.com/kibimon/monstodon/pull/3/commits/e6d38536440cd088cfbfd4af51cd7e8fb9ecf2b1
* [x] Create an `ActivityMon::Mon` model inheriting from `Account` with just additional `owner` and `species` relationships.
  * Done in https://github.com/kibimon/monstodon/pull/3/commits/65f2c89a48c545f0dd665437e9a28bb334cdf393
  * [x] Create a migration which adds `owner_id` and `species_id` columns to the `accounts` table.
    * Done in https://github.com/kibimon/monstodon/pull/3/commits/65f2c89a48c545f0dd665437e9a28bb334cdf393#diff-9a44092eadaa444aa78862d04e1b89cb
  * [x] Add a validation to `Mon` requiring that the `user` be `nil`.
    * https://github.com/kibimon/monstodon/pull/3/commits/c8bf9b4e9395625ef26cbd7a237a7baf6da6a9e5 implements this with a validation on `User`.
  * [x] Add a validation to `Mon` requiring that the `owner` must be a `Trainer`.
    * Setting the `class_name` option to `belongs_to` enforces this.
  * [x] Add the `mon` collection to `Trainer` as the inverse of `owner`.
    * Done in https://github.com/kibimon/monstodon/pull/3/commits/65f2c89a48c545f0dd665437e9a28bb334cdf393#diff-4c9c49aa598af63e3db7747eb0a72f47R53
  * [x] Add an inflection to the application configuration specifying that the plural of "mon" is "mon".
    * Done in https://github.com/kibimon/monstodon/pull/3/commits/65f2c89a48c545f0dd665437e9a28bb334cdf393#diff-cae6c97fa7e78133081d47707cf11197
* [x] Add an `Admin::ActivityMon::Species` controller
  * Done in https://github.com/kibimon/monstodon/pull/3/commits/088e770e29f147c383cf7d7ccf7362e81846dee8
  * [x] With an `index` action listing all known `Species`.
    * Done in https://github.com/kibimon/monstodon/pull/3/commits/088e770e29f147c383cf7d7ccf7362e81846dee8
  * [x] With a `new` action for a creation form.
  * [x] With a `create` action for creating new `Species`.
* [x] Add information about Mon to the `Admin::AccountsController`
* [x] Add a `create` action to the `Admin::AccountsController` that can be used for manually creating new `Mon` (eventually `Route`s too).
* [x] Add an `ActivityMon::MonController`
  * [x] With an `index` action at `/users/:username/mon` to list trainer's mon
  * [x] With a `show` action at `/users/:username/mon/:id` to view an individual mon's profile

ekiru committed:
Add ActivityMon::Species model.

ekiru committed:
Buggily add ActivityMon::Trainer

marrus-sh commented:
request: also add `summary` and `content` columns to `Species`, `Mon`

ekiru committed:
Fix naming issues caused by making things ActivityMon::Trainer

ekiru committed:
Backfill the type column for Accounts

ekiru committed:
Add ActivityMon::Mon

ekiru committed:
Validate that Users' accounts are never Mon

ekiru committed:
Add Admin::ActivityMon::SpeciesController and its index method.

ekiru committed:
Initial species creation form

marrus-sh committed:
New accounts controllers, views, validation

marrus-sh committed:
Minor fixes

ekiru commented on app/models/concerns/account_finder_concern.rb:
```diff
@@ -17,10 +17,49 @@ def find_local(username)
+    def with_id
+      search = Hash.new
+      search[id_name] = 0
+      Account.where.not(search)
```
This method can be written as just `Account.where.not(id_name => 0)

ekiru commented on app/models/concerns/account_finder_concern.rb:
```diff
@@ -17,10 +17,49 @@ def find_local(username)
 
     def find_remote(username, domain)
+      return AccountByIdFinder.new(username[4..-1], :mon_id).account if domain.nil? && username.match?(/\Amon_/i)
+      return AccountByIdFinder.new(username[6..-1], :route_no).account if domain.nil? && username.match?(/\Aroute_/i)
```
These can use `String#start_with?`.

ekiru commented on app/models/account.rb:
```diff
@@ -420,4 +454,16 @@ def normalize_domain
     self.domain = TagManager.instance.normalize_domain(domain)
   end
+
+  def not_a_mon
```
These methods should probably have a `!` in the name.

ekiru commented on app/models/concerns/account_finder_concern.rb:
```diff
@@ -17,10 +17,49 @@ def find_local(username)
+    end
+
+    def matching_id
+      Account.where(Account.arel_table[id_name].eq the_id.to_i)
```
Isn't this the same as `Account.where(id_name => the_id.to_i)`?

marrus-sh replied:
*sigh* i have no idea why we're using AREL tables here but since that's how the *other* account finder worked i tried to match the same setup

feel free to change it if it doesn't make a difference though

ekiru reviewed (commented):
just noting stuff that could be minorly tweaked. I might make the changes later

ekiru committed:
Add dex numbers and such to Species.

ekiru committed:
Add back the type column to Account

This is so single-table-inheritance works automatically.

ekiru committed:
Moves some of the user-related stuff from Account to Trainer

marrus-sh committed:
Refactored/edited Numeros

marrus-sh committed:
Show Mon Info on views

marrus-sh committed:
Fixed partial index to be <> 0 not NOT NULL

ekiru committed:
Assign trainer_no's when fabricating accounts.

ekiru committed:
Only validate presence of trainer_no if changed.

This avoids needing to reload the Trainer after creating it before
associating it with a user or such.

ekiru committed:
Better Species creation forms.

ekiru committed:
en text for activitymon species stuff

ekiru committed:
Initial species controller.

ekiru commented:
oops it occurs to me that some of my changes from tonight go beyond this PR's checklist (the bits that make a `/species/:national_no` route, which does need some fixes to get the context right and stuff)

ekiru committed:
Rename /species/:no to /Species/:no

ekiru committed:
Fix localization references for Species

ekiru committed:
Make ActivityMon::Adapter!

ekiru committed:
Specify the type when serializing Species.

marrus-sh commented:
Some things that I think we should do before we merge:

- [x] Polish up the split between different account types, routing stuff (I'm working on this; it's a big project though)
    - See f78de6335572d0d21e3d68d2770dcc6b9219fd64 which is a first attempt.
- [x] Modify namespaces
    - Change `ActivityMon` to `Monstodon` where it doesn't relate to serialization.
    - Potentially change `ActivityMon` to `MonStrPub` where it does relate to serialization, although we'll probably refactor some of this stuff in Phase 2.
    - We should probably do this last so we don't step on each other's toes.
- [x] Potentially get rid of `route_national_no` and treat route numbers the same as trainer/mon numbers
- [x] Update specs for anything that was changed in this PR
    - There will be a lot of changes to account routing, for example. (I'm working on this too.)
    - I'm not sure if we want to do specs for *everything* now, or if we just want to do specs for, e.g., v1 Trainers, and leave Mon and Routes and such for a later PR. Lmk what your thoughts are on this.
- [x] Inherit Mon stuff from Species when not otherwise specified
    - It would be easiest to just delegate `name`, etc to `species` if unset on the mon—however I'm guessing that this would also influence federation?? Probably we want to leave these blank to distinguish "Mon named Kirbi" from "Mon of species Kirbi who is unnamed"
    - So instead we'll probably have to intercept this stuff further down the chain—ie on views and such, by delegating with prefix and then using `mon.name.present? ? mon.name : mon.species_name` or so forth
    - If we wanted, we could put this in a method, like:
      ```ruby
      def computed_name
          name if name.present?
          species_name
      end
      ```
      …and then just call that, yeah?

marrus-sh committed:
Proper account splitting and routing

+ Adds a bunch of new routes for trainers/mon/routes
+ Adds routing_version to the Account model for backwards compat
+ Reworks the RoutingHelper to handle the new routes
+ Updates the various TagManagers
+ Makes use of TagManagers where possible in controllers/routes
+ Adds a new AccountAccessConcern for loading the correct account
+ Removes XML support for WebFinger since this isn't specified and I'm not sure how to make it work with the new system

This will break a bunch of specs because they haven't been updated yet.

marrus-sh committed:
A few quick fixes for the previous

marrus-sh committed:
More improvements to routing

marrus-sh committed:
Fix for AP uri_to_account

marrus-sh committed:
Spec fixes pt. 1 of ??

marrus-sh committed:
Fixed improper pluralization

ekiru committed:
Some additional spec and routing fixes.

marrus-sh commented on app/helpers/routing_helper.rb:
```diff
@@ -83,6 +83,17 @@ def default_url_options
+    when 'ActivityMon::Trainer'
+      short_trainer_path account, *more
+    end
+  end
```
Can we just add this to the previous defs so that it applies to all of the paths and not just `short_*_path`?

ekiru replied:
that would make sense

marrus-sh commented on spec/controllers/activitypub/outboxes_controller_spec.rb:
```diff
@@ -7,9 +7,23 @@
+
+  describe 'GET #show for routing version v2' do
+    before do
+      get :show, params: { account_type: 'trainer', account_no: account.trainer_no }
```
this *should* fail as-is probably because `account` is a v1 trainer so the v2 routing shouldn't work

you'll need to `Fabricate(:trainer)` in addition to the `Fabricate(:account)` to test for v2 trainers

ekiru replied:
oh that probably explains why it does fail then.

ekiru committed:
Fabricator changes for accounts and fix the outboxes controller specs.

ekiru committed:
Specify account_type param in statuses controller spec.

This fixes three of the 11 examples.

marrus-sh commented on app/serializers/activitypub/note_serializer.rb:
```diff
@@ -41,7 +41,7 @@ def published
 
   def url
-    TagManager.instance.url_for(object)
+    ActivityPub::TagManager.instance.uri_for(object)
```
Is there a reason why you made this change? `ActivityPub…uri_for` should be used for the `id` and `TagManager…url_for` should be used for the `url` in an ideal situation, I think. If it was causing a test to fail then that likely means we need to change something else……

ekiru committed:
Fix NoteSerializer#url that I messed up.

ekiru committed:
Specify account_type in RemoteFollowController specs

ekiru committed:
Specify account_type params in several account-related controller specs.

ekiru committed:
Fixes for TagManager spec and some tweaks to routing helpers.

ekiru committed:
Add missing '.instance' call on TagManger

ekiru committed:
Add in more missing `instance` calls

ekiru committed:
Fix bug in uri_to_account and update OStatus::Activity::Creation#account_from_href

ekiru committed:
Fabricate the right columns for Routes.

ekiru committed:
Use the right name for account_type in AccountAccessConcern specs.

ekiru committed:
Update AtomSerializer spec to expect the new routing.

ekiru committed:
Start numero sequences in fabricators at 1.

ekiru committed:
Update account follow/unfollow specs to expect the right redirects.

ekiru committed:
Account for slight changes in link headers in specs

ekiru committed:
Relax overly strict url regex in account_show_page_spec

ekiru committed:
Update PushController spec for changes to feed uris.

ekiru committed:
Handle stream entries in ActivityPub::TagManager#uri_for

ekiru committed:
Update specs for Account#object_type to account for account types.

ekiru committed:
Remove spec for deleted url_for method on AP::TagManager.

ekiru committed:
7 failures!!!!

ekiru committed:
Fix wrong name for url helper call

ekiru committed:
Include RoutingHelper in some places that needed its methods.

marrus-sh commented on app/controllers/accounts_controller.rb:
```diff
@@ -4,7 +4,9 @@ class AccountsController < ApplicationController
   PAGE_SIZE = 20
 
   include AccountControllerConcern
+  include RoutingHelper
```
Can we change the view for this (I'm assuming this is the problem) so that it uses the `TagManager`s and we don't need to include `RoutingHelper`?

ekiru replied:
yeah, we can do that.

ekiru replied:
(i think it might have been the controllers that inherit from it that use the helpers)

marrus-sh commented on app/controllers/auth/sessions_controller.rb:
```diff
@@ -112,7 +113,7 @@ def set_instance_presenter
     paths = [about_path]
     if single_user_mode? && resource.is_a?(User)
-      paths << short_account_path(username: resource.account)
+      paths << short_account_path(resource.account)
```
Can we just use `TagManager.instance.url_for` here so we don't have to include `RoutingHelper`?

ekiru replied:
yeah, we can do that

marrus-sh commented on app/helpers/admin/activitymon/species_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module Admin::ActivityMon::SpeciesHelper
```
Do we need this file for anything?

marrus-sh committed:
Routes now have single №; fixed MRT migrations

marrus-sh committed:
Add AS properties to Species and inherit on Mon

marrus-sh commented on app/lib/formatter.rb:
```diff
@@ -53,11 +53,11 @@ def plaintext(status)
 
   def simplified_format(account, **options)
```
Is this used by serialization?? If so, we should add an option to determine whether to use `note` (which should be used for federation) or `summary` (which should be used internally).

ekiru replied:
option added! :)

marrus-sh committed:
Updated ActivityMon -> MonStrPub

+ Now uses Monstodon:: as prefix for internal types
+ Now uses MonStrPub:: as serialization prefix
+ URIs changed to match those in latest MonStrPub spec
+ Account.[type] now uses find_no like everything else
+ Fixed errors I noticed while going through everything

ekiru committed:
simplified_format should only use summary rather than note locally

Callers that don't federate the result can pass the :allow_inheritance
option in order to use summary.

ekiru committed:
Call methods on TagManager in SessionsController & account stuff.

ekiru commented on app/controllers/accounts_controller.rb:
```diff
@@ -108,4 +117,8 @@ def filtered_status_page(params)
     end
   end
+
+  def require_account!
```
Would this make more sense on AccountAccessConcern maybe?

marrus-sh replied:
Right now (unless you've changed this) `AccountAccessConcern` is used for AP inboxes, and in the case of the shared inbox, `@account` is `nil`.

So I couldn't require the account there.

If you have a better idea for managing inboxes (maybe use a separate controller for the shared inbox that doesn't load the concern?) then feel free to move `require_account!` into there and I think it would be a good idea actually.

ekiru commented on app/controllers/home_controller.rb:
```diff
@@ -50,7 +50,7 @@ def initial_state_params
       current_account: current_account,
       token: current_session.token,
-      admin: Account.find_local(Setting.site_contact_username),
+      admin: Account.find_by_username(Setting.site_contact_username),
```
Maybe not now but eventually we should add support for identifying the site contact by trainer id (and more generally go through and make sure we don't have any legacy find_by_usernames we shouldn't).

marrus-sh replied:
I feel like this should probably be a separate PR but i'm not opposed to it

ekiru commented on app/models/account.rb:
```diff
@@ -124,20 +119,34 @@ class Account < ApplicationRecord
-
-  delegate :filtered_languages, to: :user, prefix: false, allow_nil: true
+  # Monstodon types
+  scope :mon_index, -> { where(type: 'Monstodon::Mon') }
```
I think these scopes are going to be somewhat inefficient since there's no index on type. Should we add an index on type or define these scopes in terms of the numeros since there's an index on all of those?

marrus-sh replied:
I changed the scope away from using numeros because numeros are only assigned for local entities.

IIUC, indexing only improves efficiency if the number of items with a given value is small, and it wouldn't be appropriate for something like `type`.

> Since a query searching for a common value (one that accounts for more than a few percent of all the table rows) will not use the index anyway, there is no point in keeping those rows in the index at all.
> – <https://www.postgresql.org/docs/10/static/indexes-partial.html>

ekiru replied:
ah, okay, that's fair

ekiru commented on app/views/well_known/webfinger/show.xml.ruby:
```diff
@@ -1,11 +1,11 @@
-    xml.Link(rel: 'http://webfinger.net/rel/profile-page', type: 'text/html', href: short_account_url(@account))
-    xml.Link(rel: 'http://schemas.google.com/g/2010#updates-from', type: 'application/atom+xml', href: account_url(@account, format: 'atom'))
-    xml.Link(rel: 'self', type: 'application/activity+json', href: account_url(@account))
+    xml.Alias TagManager.instance.url_for(@account)
```
We should probably make the aliases more consistent with the WebfingerSerializer for the JSON one

marrus-sh replied:
Yeah, in the case of Mon and Routes, this alias will be a duplicate of the one below.

ekiru commented on app/serializers/webfinger_serializer.rb:
```diff
@@ -10,7 +10,12 @@ def subject
   end
 
   def aliases
```
Should this use TagManager methods instead? The XML version does.

marrus-sh replied:
The XML version does because it is a View and I'm trying to keep RoutingHelper out of our controllers/views because I think TagManager gives better abstraction. But there's no reason why we couldn't use the TagManager here too.

ekiru committed:
Make webfinger aliases consistent between xrd and jrd
